### PR TITLE
PriceProvider

### PR DESCRIPTION
**Pull request by icrabbiter**
Add PriceProvider contrac

- inherit ExodiaAccessControl contract
- setTokenOracle() onlyArchitect can set token oracle address
- getCurrentPrice() returns spot market price (unsafe)
- getSafePrice() returns TWAP
- updateSafePrice() returns TWAP after updating the TWAP if the minimum amount of time has elapsed

### CODE REVIEW

File: contracts/oracles/v2/PriceProvider.sol [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    uint256 public constant VERSION = 2022030201;
+
+    mapping(address => address) public priceOracle;
+
+    /**
+     * @dev sets up the Price Oracle
+     * @param _roles exodia roles address
+     */
+    function initialize(address _roles) public initializer {
+        require(_roles != address(0), "roles cannot be null address");
```
Comment (HOVOH) [inline contracts/oracles/v2/PriceProvider.sol:24]: Let's move that requirement in the ExodiaAccessControl contract

File: test/oracles/PriceProvider.test.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+                    price.mul(2).div(100) as any
+                ); // 2% diff
+            });
+
+            it("Should be able to update safe price", async function () {
+                await priceProvider.updateSafePrice(addressRegistry.BEETS);
+            });
+        });
+
+        describe("After token oracle setting: wsSCR-FTM", function () {
```
[the remark this answers is not in the document]
Comment (icrabbiter) [reply test/oracles/PriceProvider.test.ts:206]: hmm I think it returns wsSCR price in FTM

File: test/oracles/PriceProvider.test.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+                    )
+                    .mul(xhre.ethers.utils.parseUnits("1", 18))
+                    .div(
+                        xhre.ethers.utils.parseUnits(
+                            (
+                                await getTokenPriceFromCoingecko(addressRegistry.WFTM)
+                            ).toString(),
+                            18
+                        )
+                    );
```
[the remark this answers is not in the document]
Comment (icrabbiter) [reply test/oracles/PriceProvider.test.ts:166]: you mean to check priceProvider.getCurrentPrice and balancerPriceOracle.getCurrentprice?

File: test/oracles/PriceProvider.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (HOVOH) [reply test/oracles/PriceProvider.test.ts:206]: Yes, my bad, correct the price should be in FTM. I got confused with the pool being wsSCR-DAI

File: test/oracles/PriceProvider.test.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (HOVOH) [reply test/oracles/PriceProvider.test.ts:166]: Yes, cause you technically just want to make sure the mapping is done properly as the price fetching is done and tested by the oracle already. So this test needs to validate that the proper function is called on the oracle and that it returns the same thing as the oracle call

**Pull request merged**